subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c

```diff
@@ -4,103 +4,133 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#define DT_DRV_COMPAT zephyr_ec_host_cmd_backend_espi
-
 #include <string.h>
 
 #include <zephyr/device.h>
 #include <zephyr/drivers/espi.h>
+#include <zephyr/logging/log.h>
 #include <zephyr/mgmt/ec_host_cmd/backend.h>
 #include <zephyr/mgmt/ec_host_cmd/ec_host_cmd.h>
 
-BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1, "Invalid number of eSPI backends");
+LOG_MODULE_REGISTER(host_cmd_espi, CONFIG_EC_HC_LOG_LEVEL);
 
-#define ESPI_BUS    DT_PHANDLE(DT_DRV_INST(0), bus)
-#define ESPI_DEVICE DEVICE_DT_GET(ESPI_BUS)
+#define RX_HEADER_SIZE (sizeof(struct ec_host_cmd_request_header))
 
-struct ec_host_cmd_backend_espi_data {
-	struct k_sem handler_owns;
-	struct k_sem dev_owns;
-	uint32_t rx_buffer_len;
+struct ec_host_cmd_espi_ctx {
+	/* eSPI device instance */
+	const struct device *espi_dev;
+	/* Context for read operation */
+	struct ec_host_cmd_rx_ctx *rx_ctx;
+	/* Transmit buffer */
+	struct ec_host_cmd_tx_buf *tx;
+	/* eSPI callback */
 	struct espi_callback espi_cb;
-	uint8_t *espi_shm;
 };
 
-static void ec_host_cmd_backend_espi_handler(const struct device *dev, struct espi_callback *cb,
-					    struct espi_event espi_evt)
+#define EC_HOST_CMD_ESPI_DEFINE(_name)                                                             \
+	static struct ec_host_cmd_espi_ctx _name##_hc_espi;                                        \
+	struct ec_host_cmd_backend _name = {                                                       \
+		.api = &ec_host_cmd_api,                                                           \
+		.ctx = (struct ec_host_cmd_espi_ctx *)&_name##_hc_espi,                            \
+	}
+
+static void espi_handler(const struct device *dev, struct espi_callback *cb,
+			 struct espi_event espi_evt)
 {
-	struct ec_host_cmd_backend_espi_data *data =
-		CONTAINER_OF(cb, struct ec_host_cmd_backend_espi_data, espi_cb);
+	struct ec_host_cmd_espi_ctx *hc_espi =
+		CONTAINER_OF(cb, struct ec_host_cmd_espi_ctx, espi_cb);
 	uint16_t event_type = (uint16_t)espi_evt.evt_details;
+	/* tx stores the shared memory buf pointer and size, so use it */
+	const struct ec_host_cmd_request_header *rx_header = hc_espi->tx->buf;
+	const size_t shared_size = hc_espi->tx->len_max;
+	const uint16_t rx_valid_data_size = rx_header->data_len + RX_HEADER_SIZE;
 
 	if (event_type != ESPI_PERIPHERAL_EC_HOST_CMD) {
 		return;
 	}
 
-	if (k_sem_take(&data->dev_owns, K_NO_WAIT) != 0) {
-		int res = EC_HOST_CMD_IN_PROGRESS;
-
-		espi_write_lpc_request(ESPI_DEVICE, ECUSTOM_HOST_CMD_SEND_RESULT, &res);
-		return;
+	/* Only support version 3 and make sure the number of bytes to copy is not
+	 * bigger than rx buf size or the shared memory size
+	 */
+	if (rx_header->prtcl_ver != 3 ||
+	    rx_valid_data_size > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER ||
+	    rx_valid_data_size > shared_size) {
+		memcpy(hc_espi->rx_ctx->buf, (void *)rx_header, RX_HEADER_SIZE);
```

Comment:
Don't you need a `k_sem_take()` before writing into the RX buffer?  Can the host send another command while the handler is still processing a command?

Reply:
Done.